lib/Runtime/Library/JavascriptBigInt.cpp

```diff
@@ -397,4 +414,183 @@ namespace Js
+        // adjust length
+        while ((pbi1->m_length>0) && (pbi1->m_digits[pbi1->m_length-1] == 0))
+        {
+            pbi1->m_length--;
+        }
+        Assert(pbi1->m_length > 0);
+    }
+
+    JavascriptBigInt * JavascriptBigInt::Sub(JavascriptBigInt * pbi1, JavascriptBigInt * pbi2)
+    {
+        if (JavascriptBigInt::IsZero(pbi1))
+        {
+            pbi2->m_isNegative = !pbi2->m_isNegative;
+            return pbi2;
+        }
+        if (JavascriptBigInt::IsZero(pbi2))
+        {
+            return pbi1;
+        }
+
+        if (pbi2->m_isNegative)
+        {
+            pbi2->m_isNegative = false;
+            return JavascriptBigInt::Add(pbi1, pbi2);// a-(-b)=a+b
+        }
+
+        if (pbi1->m_isNegative)
+        {
+            JavascriptBigInt::AddAbsolute(pbi1, pbi2);// -a-b=-(a+b)
+            return pbi1;
+        }
+        else // both positive 
+        {
+            if (pbi1->CompareAbsolute(pbi2) == 0) // a -a = 0
+            {
+                return JavascriptBigInt::CreateZero(pbi1->GetScriptContext());
+            }
+            else if (pbi1->CompareAbsolute(pbi2) == 1) 
+            {
+                JavascriptBigInt::SubAbsolute(pbi1, pbi2); // a - b > 0
+                return pbi1;
+            }
+            else
+            {
+                pbi2->m_isNegative = true;
+                JavascriptBigInt::SubAbsolute(pbi2, pbi1); // a - b = - (b-a) < 0
+                return pbi2;
+            }
+        }
+    }
+
+    JavascriptBigInt * JavascriptBigInt::Add(JavascriptBigInt * pbi1, JavascriptBigInt * pbi2)
+    {
+        if (JavascriptBigInt::IsZero(pbi1)) 
+        {
+            return pbi2;
+        }
+        if (JavascriptBigInt::IsZero(pbi2))
+        {
+            return pbi1;
+        }
+
+        if (pbi1->m_isNegative == pbi2->m_isNegative)
+        {
+            JavascriptBigInt::AddAbsolute(pbi1, pbi2); // (-a)+(-b) = -(a+b)
+        }
+        else 
+        {
+            if (pbi1->CompareAbsolute(pbi2) == 0)
+            {
+                return JavascriptBigInt::CreateZero(pbi1->GetScriptContext()); // a + (-a) = -a + a = 0
+            }
+            else if (pbi1->CompareAbsolute(pbi2) == 1) 
+            {
+                 JavascriptBigInt::SubAbsolute(pbi1, pbi2); // a + (-b) = a - b or (-a) + b = -(a-b)
+                 return pbi1;
+            }
+            else
+            {
+                JavascriptBigInt::SubAbsolute(pbi2, pbi1); // -a + b = b - a or a + (-b) = -(b-a)
+                return pbi2;
+            }
+        }
+        return pbi1;
+    }
+
+    Var JavascriptBigInt::Add(Var aLeft, Var aRight)
+    {
+        JavascriptBigInt *leftBigInt = VarTo<JavascriptBigInt>(aLeft);
+        JavascriptBigInt *rightBigInt = VarTo<JavascriptBigInt>(aRight);
+        return JavascriptBigInt::Add(JavascriptBigInt::New(leftBigInt, leftBigInt->GetScriptContext()), JavascriptBigInt::New(rightBigInt, rightBigInt->GetScriptContext()));
```

Comment:
Currently this does two allocations, and ideally we would need only one to create the output BigInt (or zero in the special case where one of the args is zero, because we can return the other arg directly by reference). I don't think we're worried about fine-tuning perf at this point, so this is fine, but maybe it would be worth adding a comment with something like "Consider deferring creation of new instances until we need them".

Reply:
will add TODO